src/lib/operators/table_scan/sorted_segment_search.hpp

```diff
@@ -0,0 +1,156 @@
+   * sort order _get_first_bound will actually return an upper bound and _get_last_bound the lower one. However, the
+   * first offset will always point to an entry matching the search value, whereas last offset points to the entry
+   * behind the last matching one.
+   */
+  IteratorType _get_first_bound() const {
+    if (_is_ascending) {
+      return std::lower_bound(_begin, _end, _search_value, [](const auto& segment_position, const auto& search_value) {
+        return segment_position.value() < search_value;
+      });
+    } else {
+      return std::lower_bound(_begin, _end, _search_value, [](const auto& segment_position, const auto& search_value) {
+        return segment_position.value() > search_value;
+      });
+    }
+  }
+
+  IteratorType _get_last_bound() const {
+    if (_is_ascending) {
+      return std::upper_bound(_begin, _end, _search_value, [](const auto& search_value, const auto& segment_position) {
+        return segment_position.value() > search_value;
+      });
+    } else {
+      return std::upper_bound(_begin, _end, _search_value, [](const auto& search_value, const auto& segment_position) {
+        return segment_position.value() < search_value;
+      });
+    }
+  }
+
+  // This function sets the offset(s) which delimit the result set based on the predicate condition and the sort order
+  void _set_begin_and_end() {
+    if (_predicate_condition == PredicateCondition::Equals) {
+      _begin = _get_first_bound();
+      _end = _get_last_bound();
+      return;
+    }
+
+    // clang-format off
+    if (_is_ascending) {
+      switch (_predicate_condition) {
+        case PredicateCondition::GreaterThanEquals: _begin = _get_first_bound(); return;
+        case PredicateCondition::GreaterThan: _begin = _get_last_bound(); return;
+        case PredicateCondition::LessThanEquals: _end = _get_last_bound(); return;
+        case PredicateCondition::LessThan: _end = _get_first_bound(); return;
+        default: Fail("Unsupported predicate condition encountered");
+      }
+    } else {
+      switch (_predicate_condition) {
+        case PredicateCondition::LessThanEquals: _begin = _get_first_bound(); return;
+        case PredicateCondition::LessThan: _begin = _get_last_bound(); return;
+        case PredicateCondition::GreaterThanEquals: _end = _get_last_bound(); return;
+        case PredicateCondition::GreaterThan: _end = _get_first_bound(); return;
+        default: Fail("Unsupported predicate condition encountered");
+      }
+    }
+    // clang-format on
+  }
+
+  /*
+   * NotEquals might result in two matching ranges (one below and one above the search_value) and needs special handling
+   * Two criteria for early outs can be when there is only one matching range:
+   * 1. The search value is not present -> matches everything
+   * 2. The search value is present but only at the end of the range -> matches until first occurrence of search value
```

Comment:
what about "The search value is present but only at the **beginning** of the range"?

Reply:
I have rewritten the comment and added the missing `if` for that case. FYI: These early-outs are only for performance reasons and don't change the functionality. Further details are in the new comments.